### ISSUE TITLE
Auth+Ticket+Ticket Validation Example?

### ISSUE THREAD

**Issue by dylanh724** [Facepunch/Facepunch.Steamworks#207]
## EDIT
Skip to bottom -- in between is progression.

## Original Issue

I could only find `Auth.GetAuthSessionTicket` that returns a `byte[]` via `_ticket.data`:

What do I do from there? In Steamworks.net, it's stupidly complicated to convert it, then you need to send it away for a callback:
```csharp
ticketBlob = new byte[1024];

// Retrieve ticket; hTicket should be a field in the class so you can use it to cancel the ticket later
// When you pass an object, the object can be modified by the callee. This function modifies the byte array you've passed to it.
hTicket = SteamUser.GetAuthSessionTicket(ticketBlob, ticketBlob.Length, out ticketSize);

// Resize the buffer to actual length
Array.Resize(ref ticketBlob, (int)ticketSize);

// Convert bytes to string
StringBuilder sb = new StringBuilder();
foreach (byte b in ticketBlob)
{
	sb.AppendFormat("{0:x2}", b);
}
hexTicket = sb.ToString(); // Final result: Then you send it off to Steam to get verified. Send this hexTicket to your BaaS or oAuth2.

// Wait for OnGetAuthSessionTicketResponse to confirm the ticket is valid =>

private void OnGetAuthSessionTicketResponse(GetAuthSessionTicketResponse_t callback)
{
    // Confirm validity of ticket here
}
```
How do I do the Facepunch equiv of this? Rather:

1. How do you convert an `Auth.Ticket` to string (the final form)?
2. How do I validate the ticket (`OnGetAuthSessionTicketResponse()` equiv.)?

There used to be an example link via this issue ( https://github.com/Facepunch/Facepunch.Steamworks/issues/52 ). However, it's now 404.

**Comment by dylanh724** [comment]
Think I got it:
```csharp
using Facepunch.Steamworks;

void auth()
{
	ulong steamId64 = Client.Instance.SteamId;
	Auth.Ticket ticket = s_client.Auth.GetAuthSessionTicket();
	Server.Instance.Auth.StartSession(ticket.Data, SteamInfo.steamId64);
	Server.Instance.Auth.OnAuthChange += OnAuthChange;
}

void OnAuthChange(ulong steamId, ulong ownerId, ServerAuth.Status status)
{
	bool isSuccess = (status == ServerAuth.Status.OK);
}
```

**Comment by dylanh724** [comment]
Gonna update this for TL;DR above -- apologies for my post. I'm trying to get this rolled out and so close:

##TL;DR:
Seems that server isn't just for game servers -- I currently have almost everything running *except* `ServerInit` is not a valid <type> for `ServerInit options` from src: What's the new way to init server?

**Comment by dylanh724** [comment]
^ That's just progression -- ignore that.

Ok, so everything seems to work EXCEPT the auth callback:

```
void OnEnable()
{
    Server.Instance.Auth.OnAuthChange += OnAuthChange;
}

void startAuth()
{
   authWorksHaveTicketTriggerCallback();
}

void OnAuthChange(ulong steamId, ulong ownerId, ServerAuth.Status status)
{
    // CALLBACK NEVER GETS TRIGGERED: How come?
}
```

**Comment by igniuss** [comment]
If you want to verify the session ticket in a single player experience you'd need to post it to the `AuthenticateUserTicket` web-api of steamworks, not `Server.Instance`

The other method would be having a dedicated server that initializes a Server instance 
example: `var server = new Facepunch.Steamworks.Server(252490, options);` (see the Facepunch.Test assembly for a big example)

After which you can use the `StartSession` part on the server, and send a callback to the client if it was correct or not.

I haven't played around with it yet but, 
```csharp
byte[] ticketBlob = authTicket.Data;
string s = string.Concat(ticketBlob.Select(b => b.ToString("X2")));
```

seems correct, should just post it to the webapi (see [the Steam WebAPI doc for more info](https://partner.steamgames.com/doc/webapi/ISteamUserAuth) )

You can also [See this Test-code to see it's working](https://github.com/Facepunch/Facepunch.Steamworks/blob/438127288a47b3c7032cb7783d389ee779e76b5e/Facepunch.Steamworks.Test/Client/Server/ServerTest.cs#L63)
only difference is you're testing it locally instead of sending over the binary to the server who then checks it.

**Comment by dylanh724** [comment]
Oh man, I've been doing it wrong this whole time, then - even since steamworks.net, perhaps. I always thought that authenticating changes the ticket ... but apparently the ticket is just arbitrary? Testing ;D

**Comment by dylanh724** [comment]
^ Indeed, I didn't need that! Got it working~

**Comment by dylanh724** [comment]
Alas, it seems that 50%+ of the time, users are getting an "Invalid Ticket:

`{ \"response\" : { \"error\" : { \"errorcode\" : 101 , \"errordesc\" : \"Invalid ticket\"}}}`

Anyone know what would cause this?

Related: https://github.com/Facepunch/Facepunch.Steamworks/issues/217
